### ISSUE TITLE
Compiler crash with optional closure (possibly Observable related)

### ISSUE THREAD

**Issue by jasonbobier** [apple/swift#67324]
**Description**

The following code crashes:

```
import Foundation
import Observation

@Observable
class ObservedClass {
	var name: String
	var closure: (() -> Void)?

	init(name: String, closure: (() -> Void)? = nil) {
		self.name = name
		self.closure = closure
	}
}
```

**Environment**

Xcode: Version 15.0 beta 4 (15A5195m)

iOS 17.0 beta 4

```
Stack dump:
0.	Program arguments: /Applications/Xcode-beta.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/swift-frontend -frontend -c -primary-file /Users/jason/Desktop/Test/Test/ObservedClass.swift /Users/jason/Desktop/Test/Test/ContentView.swift /Users/jason/Desktop/Test/Test/TestApp.swift /Users/jason/Library/Developer/Xcode/DerivedData/Test-ghgwcntlalxbpmeuaekqwfcdbrxm/Build/Intermediates.noindex/Test.build/Debug/Test.build/DerivedSources/GeneratedAssetSymbols.swift -emit-dependencies-path /Users/jason/Library/Developer/Xcode/DerivedData/Test-ghgwcntlalxbpmeuaekqwfcdbrxm/Build/Intermediates.noindex/Test.build/Debug/Test.build/Objects-normal/arm64/ObservedClass.d -emit-const-values-path /Users/jason/Library/Developer/Xcode/DerivedData/Test-ghgwcntlalxbpmeuaekqwfcdbrxm/Build/Intermediates.noindex/Test.build/Debug/Test.build/Objects-normal/arm64/ObservedClass.swiftconstvalues -emit-reference-dependencies-path /Users/jason/Library/Developer/Xcode/DerivedData/Test-ghgwcntlalxbpmeuaekqwfcdbrxm/Build/Intermediates.noindex/Test.build/Debug/Test.build/Objects-normal/arm64/ObservedClass.swiftdeps -serialize-diagnostics-path /Users/jason/Library/Developer/Xcode/DerivedData/Test-ghgwcntlalxbpmeuaekqwfcdbrxm/Build/Intermediates.noindex/Test.build/Debug/Test.build/Objects-normal/arm64/ObservedClass.dia -emit-localized-strings -emit-localized-strings-path /Users/jason/Library/Developer/Xcode/DerivedData/Test-ghgwcntlalxbpmeuaekqwfcdbrxm/Build/Intermediates.noindex/Test.build/Debug/Test.build/Objects-normal/arm64 -target arm64-apple-macos14.0 -Xllvm -aarch64-use-tbi -enable-objc-interop -stack-check -sdk /Applications/Xcode-beta.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX14.0.sdk -I /Users/jason/Library/Developer/Xcode/DerivedData/Test-ghgwcntlalxbpmeuaekqwfcdbrxm/Build/Products/Debug -F /Users/jason/Library/Developer/Xcode/DerivedData/Test-ghgwcntlalxbpmeuaekqwfcdbrxm/Build/Products/Debug -no-color-diagnostics -enable-testing -g -module-cache-path /Users/jason/Library/Developer/Xcode/DerivedData/ModuleCache.noindex -swift-version 5 -enforce-exclusivity=checked -Onone -D DEBUG -serialize-debugging-options -const-gather-protocols-file /Users/jason/Library/Developer/Xcode/DerivedData/Test-ghgwcntlalxbpmeuaekqwfcdbrxm/Build/Intermediates.noindex/Test.build/Debug/Test.build/Objects-normal/arm64/Test_const_extract_protocols.json -enable-bare-slash-regex -empty-abi-descriptor -plugin-path /Applications/Xcode-beta.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/lib/swift/host/plugins -plugin-path /Applications/Xcode-beta.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/local/lib/swift/host/plugins -validate-clang-modules-once -clang-build-session-file /Users/jason/Library/Developer/Xcode/DerivedData/ModuleCache.noindex/Session.modulevalidation -Xcc -working-directory -Xcc /Users/jason/Desktop/Test -resource-dir /Applications/Xcode-beta.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/lib/swift -enable-anonymous-context-mangled-names -Xcc -ivfsstatcache -Xcc /Users/jason/Library/Developer/Xcode/DerivedData/SDKStatCaches.noindex/macosx14.0-23A5286f-f7e79a2b09dbe2c8a4ba251adb57f722.sdkstatcache -Xcc -I/Users/jason/Library/Developer/Xcode/DerivedData/Test-ghgwcntlalxbpmeuaekqwfcdbrxm/Build/Intermediates.noindex/Test.build/Debug/Test.build/swift-overrides.hmap -Xcc -iquote -Xcc /Users/jason/Library/Developer/Xcode/DerivedData/Test-ghgwcntlalxbpmeuaekqwfcdbrxm/Build/Intermediates.noindex/Test.build/Debug/Test.build/Test-generated-files.hmap -Xcc -I/Users/jason/Library/Developer/Xcode/DerivedData/Test-ghgwcntlalxbpmeuaekqwfcdbrxm/Build/Intermediates.noindex/Test.build/Debug/Test.build/Test-own-target-headers.hmap -Xcc -I/Users/jason/Library/Developer/Xcode/DerivedData/Test-ghgwcntlalxbpmeuaekqwfcdbrxm/Build/Intermediates.noindex/Test.build/Debug/Test.build/Test-all-target-headers.hmap -Xcc -iquote -Xcc /Users/jason/Library/Developer/Xcode/DerivedData/Test-ghgwcntlalxbpmeuaekqwfcdbrxm/Build/Intermediates.noindex/Test.build/Debug/Test.build/Test-project-headers.hmap -Xcc -I/Users/jason/Library/Developer/Xcode/DerivedData/Test-ghgwcntlalxbpmeuaekqwfcdbrxm/Build/Products/Debug/include -Xcc -I/Users/jason/Library/Developer/Xcode/DerivedData/Test-ghgwcntlalxbpmeuaekqwfcdbrxm/Build/Intermediates.noindex/Test.build/Debug/Test.build/DerivedSources-normal/arm64 -Xcc -I/Users/jason/Library/Developer/Xcode/DerivedData/Test-ghgwcntlalxbpmeuaekqwfcdbrxm/Build/Intermediates.noindex/Test.build/Debug/Test.build/DerivedSources/arm64 -Xcc -I/Users/jason/Library/Developer/Xcode/DerivedData/Test-ghgwcntlalxbpmeuaekqwfcdbrxm/Build/Intermediates.noindex/Test.build/Debug/Test.build/DerivedSources -Xcc -DDEBUG=1 -module-name Test -frontend-parseable-output -disable-clang-spi -target-sdk-version 14.0 -target-sdk-name macosx14.0 -o /Users/jason/Library/Developer/Xcode/DerivedData/Test-ghgwcntlalxbpmeuaekqwfcdbrxm/Build/Intermediates.noindex/Test.build/Debug/Test.build/Objects-normal/arm64/ObservedClass.o -index-unit-output-path /Test.build/Debug/Test.build/Objects-normal/arm64/ObservedClass.o -index-store-path /Users/jason/Library/Developer/Xcode/DerivedData/Test-ghgwcntlalxbpmeuaekqwfcdbrxm/Index.noindex/DataStore -index-system-modules
1.	Apple Swift version 5.9 (swiftlang-5.9.0.120.7 clang-1500.0.34.3)
2.	Compiling with the current language version
3.	While evaluating request IRGenRequest(IR Generation for file "/Users/jason/Desktop/Test/Test/ObservedClass.swift")
4.	While evaluating request ExecuteSILPipelineRequest(Run pipelines { IRGen Preparation } on SIL for Test)
5.	While running pass #31 SILModuleTransform "LoadableByAddress".
6.	While converting type 'Optional<() -> ()>'
7.	While converting type '() -> ()'
Stack dump without symbol names (ensure you have llvm-symbolizer in your PATH or set the environment var `LLVM_SYMBOLIZER_PATH` to point to it):
0  swift-frontend           0x00000001053d8a4c llvm::sys::PrintStackTrace(llvm::raw_ostream&, int) + 56
1  swift-frontend           0x00000001053d7f08 llvm::sys::RunSignalHandlers() + 112
2  swift-frontend           0x0000000103c4b494 SignalHandler(int) + 352
3  libsystem_platform.dylib 0x00000001a2d9ea24 _sigtramp + 56
4  swift-frontend           0x0000000103793064 swift::irgen::TypeConverter::getTypeEntry(swift::CanType) + 1008
5  swift-frontend           0x00000001054707b4 swift::irgen::EnumImplStrategy::get(swift::irgen::TypeConverter&, swift::SILType, swift::EnumDecl*) + 2748
6  swift-frontend           0x000000010547a94c swift::irgen::TypeConverter::convertEnumType(swift::TypeBase*, swift::CanType, swift::EnumDecl*) + 356
7  swift-frontend           0x0000000106a2bbc4 swift::irgen::TypeConverter::convertAnyNominalType(swift::CanType, swift::NominalTypeDecl*) + 1168
8  swift-frontend           0x00000001037930c8 swift::irgen::TypeConverter::getTypeEntry(swift::CanType) + 1108
9  swift-frontend           0x00000001077faab0 isLargeLoadableType(swift::GenericEnvironment*, swift::SILType, swift::irgen::IRGenModule&) + 668
10 swift-frontend           0x00000001077f3e94 LargeSILTypeMapper::getNewSILType(swift::GenericEnvironment*, swift::SILType, swift::irgen::IRGenModule&) + 952
11 swift-frontend           0x0000000103ac52a4 (anonymous namespace)::LoadableByAddress::runOnFunction(swift::SILFunction*) + 828
12 swift-frontend           0x000000010509ed84 (anonymous namespace)::LoadableByAddress::run() + 84
13 swift-frontend           0x000000010777a454 swift::SILPassManager::executePassPipelinePlan(swift::SILPassPipelinePlan const&) + 16212
14 swift-frontend           0x00000001078f8ce8 swift::SimpleRequest<swift::ExecuteSILPipelineRequest, std::__1::tuple<> (swift::SILPipelineExecutionDescriptor), (swift::RequestFlags)1>::evaluateRequest(swift::ExecuteSILPipelineRequest const&, swift::Evaluator&) + 56
15 swift-frontend           0x00000001077b5774 llvm::Expected<swift::ExecuteSILPipelineRequest::OutputType> swift::Evaluator::getResultUncached<swift::ExecuteSILPipelineRequest>(swift::ExecuteSILPipelineRequest const&) + 476
16 swift-frontend           0x0000000104b0715c swift::IRGenRequest::evaluate(swift::Evaluator&, swift::IRGenDescriptor) const + 6340
17 swift-frontend           0x00000001074cafe4 swift::SimpleRequest<swift::IRGenRequest, swift::GeneratedModule (swift::IRGenDescriptor), (swift::RequestFlags)9>::evaluateRequest(swift::IRGenRequest const&, swift::Evaluator&) + 176
18 swift-frontend           0x0000000106f746d8 llvm::Expected<swift::IRGenRequest::OutputType> swift::Evaluator::getResultUncached<swift::IRGenRequest>(swift::IRGenRequest const&) + 1480
19 swift-frontend           0x0000000106ce71d0 swift::performIRGeneration(swift::FileUnit*, swift::IRGenOptions const&, swift::TBDGenOptions const&, std::__1::unique_ptr<swift::SILModule, std::__1::default_delete<swift::SILModule>>, llvm::StringRef, swift::PrimarySpecificPaths const&, llvm::StringRef, llvm::GlobalVariable**) + 264
20 swift-frontend           0x0000000107dc4154 generateIR(swift::IRGenOptions const&, swift::TBDGenOptions const&, std::__1::unique_ptr<swift::SILModule, std::__1::default_delete<swift::SILModule>>, swift::PrimarySpecificPaths const&, llvm::StringRef, llvm::PointerUnion<swift::ModuleDecl*, swift::SourceFile*>, llvm::GlobalVariable*&, llvm::ArrayRef<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>>) + 156
21 swift-frontend           0x0000000107dcd03c performCompileStepsPostSILGen(swift::CompilerInstance&, std::__1::unique_ptr<swift::SILModule, std::__1::default_delete<swift::SILModule>>, llvm::PointerUnion<swift::ModuleDecl*, swift::SourceFile*>, swift::PrimarySpecificPaths const&, int&, swift::FrontendObserver*) + 1596
22 swift-frontend           0x0000000107dc8234 performCompile(swift::CompilerInstance&, int&, swift::FrontendObserver*) + 1748
23 swift-frontend           0x0000000107dcbba0 swift::performFrontend(llvm::ArrayRef<char const*>, char const*, void*, swift::FrontendObserver*) + 4216
24 swift-frontend           0x0000000107727830 swift::mainEntry(int, char const**) + 4112
25 dyld                     0x00000001a2a17f28 start + 2236
```

**Comment by xedin** [comment]
I have a fix for this in https://github.com/apple/swift/pull/67312

**Comment by jasonbobier** [comment]
This appears to be fixed in 5.9.
